File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.js
function solution(numbers, hand) {
    let left = "*";
    let right = "#";
    let result = "";
    
    const getDistance = (locatedNumber, target) => {
        const keyPad = {
            1: [0, 0],
            2: [0, 1],
            3: [0, 2],
            4: [1, 0],
            5: [1, 1],
            6: [1, 2],
            7: [2, 0],
            8: [2, 1],
            9: [2, 2],
            "*": [3, 0],
            0: [3, 1],
            "#": [3, 2],
        };
        
        const nowPosition = keyPad[locatedNumber];
        const targetPosition = keyPad[target];
        
        return (
            Math.abs(targetPosition[0] - nowPosition[0])
            + Math.abs(targetPosition[1] - nowPosition[1])
        );
    };
    
    for(let i = 0; i < numbers.length; i++) {
        if((numbers[i] == 1) || (numbers[i] == 4) || (numbers[i] == 7)){
            result += "L";
            left = numbers[i];
        } else if((numbers[i] == 3) || (numbers[i] == 6) || (numbers[i] == 9)) {
            result += "R"
            right = numbers[i];
        } else {
            let leftDistance = getDistance(left, numbers[i]);
            let rightDistance = getDistance(right, numbers[i]);

            if(leftDistance < rightDistance) {
                result += "L";
                left = numbers[i];
            } else if(leftDistance == rightDistance) {
                if(hand == "left") {
                    result += "L";
                    left = numbers[i];
                } else {
                    result += "R"
                    right = numbers[i];
                }
            } else {
                result += "R"
                right = numbers[i];
            }
        }
    }
    return result;
}

File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.py
def solution(numbers, hand):
    result = ""
    left = "*"
    right = "#"
    
    def get_distance(located_num, target):
        keypad = {
            1: [0, 0],
            2: [0, 1],
            3: [0, 2],
            4: [1, 0],
            5: [1, 1],
            6: [1, 2],
            7: [2, 0],
            8: [2, 1],
            9: [2, 2],
            "*": [3, 0],
            0: [3, 1],
            "#": [3, 2],
        }
        
        now_position = keypad[located_num]
        target_position = keypad[target]
        
        return (
            abs(target_position[0] - now_position[0])
            + abs(target_position[1] - now_position[1])
        )
        
    for num in numbers:
        if((num == 1) or (num == 4) or (num == 7)):
            result += "L"
            left = num
        elif((num == 3) or (num == 6) or (num == 9)):
            result += "R"
            right = num
        else:
            left_distance = get_distance(left, num)
            right_distance = get_distance(right, num)
            
            if(left_distance < right_distance):
                result += "L"
                left = num
            elif(left_distance == right_distance):
                if hand == "left":
                    result += "L"
                    left = num
                else:
                    result += "R"
                    right = num
            else:
                result += "R"
                right = num
                
    return result